packages/git/src/node/git-repository-locator.ts

```diff
@@ -4,32 +4,75 @@
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 */
+
+import { injectable } from 'inversify';
 import * as fs from 'fs';
-import * as Path from 'path';
+import * as paths from 'path';
 import { FileUri } from '@theia/core/lib/node/file-uri';
 import { Repository } from '../common/model';
 
+const ignore: () => IgnoreFilter = require('ignore');
 const abs: (path: string) => string = require('abs');
-const finder: any = require('findit2');
-
-/**
- * Resolves to an array of repositories, recursively discovered from the given root `path`.
- *
- * @param path the FS path of the root to start the discovery.
- */
-export function locateRepositories(path: string): Promise<Repository[]> {
-    return new Promise<Repository[]>((resolve, reject) => {
-        const repositoryPaths = new Set();
-        const emitter = finder(abs(path));
-        emitter.on('directory', (dir: string, stat: fs.Stats, stop: () => void) => {
-            const base = Path.basename(dir);
-            if (base === '.git') {
-                const dirName = Path.dirname(dir);
-                repositoryPaths.add(dirName);
-                stop();
-            }
+const finder = require('findit2');
+
+export interface IgnoreFilter {
+    add(patterns: string): void
+    ignores(pathname: string): boolean;
+}
+export type IngoreFilterFn = (pathname: string) => boolean;
+
+@injectable()
+export class GitRepositoryLocator {
+
+    /**
+     * Resolves to an array of repositories, recursively discovered from the given root `path`.
+     *
+     * @param path the FS path of the root to start the discovery.
+     */
+    locate(path: string): Promise<Repository[]> {
+        return new Promise<Repository[]>((resolve, reject) => {
+            const filters = new Map<string, (pathname: string) => boolean>();
+            const repositoryPaths = new Set();
+            const emitter = finder(abs(path));
+            emitter.on('directory', (dir: string, stat: fs.Stats, stop: () => void) => {
+                const base = paths.basename(dir);
+                if (base === '.git') {
+                    const dirName = paths.dirname(dir);
+                    repositoryPaths.add(dirName);
+                    stop();
+                    return;
+                }
+                if (this.shouldStop(dir, filters)) {
```

Comment:
the filter should be applied first, no?

Reply:
You think someone will include `.git` in `.gitignore`?

Reply:
sorry, my bad. I didn't understand the code correctly at first